tools/distpartitioning/multi_dev_init.py

```diff
@@ -0,0 +1,499 @@
+
+
+def exchange_node_features(rank, world_size, node_data, node_features, ntypes_map, \
+        ntypes_nid_map, ntype_id_count, node_part_ids):
+    """
+    This function is used to shuffle node features so that each process will receive
+    all the node features whose corresponding nodes are owned by the same process. 
+    The mapping procedure to identify the owner process is not straight forward. The
+    following steps are used to identify the owner processes for the locally read node-
+    features. 
+    a. Compute the global_nids for the locally read node features. Here metadata json file
+        is used to identify the corresponding global_nids. Please note that initial graph input
+        nodes.txt files are sorted based on node_types. 
+    b. Using global_nids and metis partitions owner processes can be easily identified. 
+    c. Now each process sends the global_nids for which shuffle_global_nids are needed to be 
+        retrieved. 
+    d. After receiving the corresponding shuffle_global_nids these ids are added to the 
+        node_data and edge_data dictionaries
+
+    Parameters: 
+    -----------
+    rank : int
+        rank of the current process
+    world_size : int
+        total no. of participating processes. 
+    node_data : dictionary
+        dictionary where node data is stored, which is initially read from the nodes txt file mapped
+        to the current process
+    node_feautres: dicitonary
+        dictionry where node_features are stored and this information is read from the appropriate
+        node features file which belongs to the current process
+    ntypes_map : dictionary
+        mappings between node type names and node type ids
+    ntypes_nid_map : dictionary
+        mapping between node type names and global_nids which belong to the keys in this dictionary
+    ntype_id_count : dictionary
+        mapping between node type id and no of nodes which belong to each node_type_id
+    node_part_ids : numpy array
+        numpy array which store the partition-ids and indexed by global_nids
+    """
+
+    #determine Global_type_nid for the residing features 
+    start = timer()
+    node_features_rank_lst = []
+    global_nid_rank_lst = []
+    for part_id in np.arange(world_size):
+
+        #form outgoing features to each process
+        send_node_features = {}
+        send_global_nids = {}
+        for ntype_name, ntype_id in ntypes_map.items(): 
+
+            #check if features exist for this node_type
+            if (ntype_name+'/feat' in node_features) and (node_features[ntype_name+'/feat'].shape[0] > 0):
+                feature_count = node_features[ntype_name+'/feat'].shape[0]
+                global_feature_count = ntype_id_count[str(ntype_id)]
+
+                #determine the starting global_nid for this node_type_id
+                feat_per_proc = math.ceil(global_feature_count / world_size)
+                global_type_nid_start = feat_per_proc * rank
```

Comment:
i don't quite understand what we try to do here. `global_type_nid_start` and `global_type_nid_end` indicates the per-type node id range for the node features read from the disk? if so, why can we calculate this? do you assume how the node features are partitioned?

Reply:
Here the process is simple. node-features are mapped to appopriate owner processes in the following steps: 
1. node features are read. 
2. global_type_nid_start and global_type_nid_end are calculated 
3. global_nid for the node features read from file are calculated 
4. with the help of global_nids, we use the node_id to partition_id map to get the owner process
5. We do the above steps for all the ranks in the world.
6. when building node_features_rank_list and global_nid_rank_list, we perform alltoallv and send/receive the node features with corresponding global_nids.